framework/src/modules/chain/bft/finality_manager.js

```diff
@@ -0,0 +1,297 @@
+		assert(activeDelegates > 0, 'Must provide a positive activeDelegates');
+
+		// Set constants
+		this.activeDelegates = activeDelegates;
+
+		// Threshold to consider a block pre-voted
+		this.preVoteThreshold = Math.ceil((this.activeDelegates * 2) / 3);
+
+		// Threshold to consider a block pre-committed (or finalized)
+		this.preCommitThreshold = Math.ceil((this.activeDelegates * 2) / 3);
+
+		// Limit for blocks to make perform verification or pre-vote/pre-commit (1 block less than 3 rounds)
+		this.processingThreshold = this.activeDelegates * 3 - 1;
+
+		// Maximum headers to store (5 rounds)
+		this.maxHeaders = this.activeDelegates * 5;
+
+		this.headers = new HeadersList({ size: this.maxHeaders });
+
+		// Height up to which blocks are finalized
+		this.finalizedHeight = finalizedHeight;
+
+		// Height up to which blocks have pre-voted
+		this.prevotedConfirmedHeight = 0;
+
+		this.state = {};
+		this.preVotes = {};
+		this.preCommits = {};
+	}
+
+	/**
+	 * Add block header to BlockHeaderManager
+	 *
+	 * @param {BlockHeader} blockHeader
+	 * @return {Block_headers_manager}
+	 */
+	addBlockHeader(blockHeader) {
+		debug('addBlockHeader invoked');
+		debug('validateBlockHeader invoked');
+		// Validate the schema of the header
+		// To spy exported function in same module we have to call it as this
+		validateBlockHeader(blockHeader);
+
+		// Verify the integrity of the header with chain
+		try {
+			this.verifyBlockHeaders(blockHeader);
+		} catch (error) {
+			// TODO: Remove hardcoded value of maxHeightPreviouslyForged to avoid this
+			// https://github.com/LiskHQ/lisk-sdk/blob/fa1bb6907955c12297336f80f59951ba4754da7f/framework/src/modules/chain/blocks/process.js#L125-L126
+			if (!(error instanceof BFTChainDisjointError)) {
+				throw error;
+			}
+		}
+
+		// Add the header to the list
+		this.headers.add(blockHeader);
+
+		// Update the pre-votes and pre-commits
+		this.updatePreVotesPreCommits(blockHeader);
+
+		// Update the pre-voted confirmed and finalized height
+		this.updatePreVotedAndFinalizedHeight();
+
+		debug('after adding block header', {
+			finalizedHeight: this.finalizedHeight,
+			prevotedConfirmedHeight: this.prevotedConfirmedHeight,
+			minHeight: this.minHeight,
+			maxHeight: this.maxHeight,
+		});
+		return this;
+	}
+
+	/**
+	 * Update pre-votes and pre-commits in reference to particular block header
+	 *
+	 * @param {BlockHeader} lastBlockHeader
+	 * @return {undefined}
+	 */
+	updatePreVotesPreCommits(lastBlockHeader) {
+		debug('updatePreVotesPreCommits invoked');
+		// Update applies particularly in reference to last block header in the list
+		const header = lastBlockHeader || this.headers.last;
+
+		// If delegate forged a block with higher or same height previously
+		// that means he is forging on other chain and we don't count any
+		// pre-votes and pre-commits from him
+		if (header.maxHeightPreviouslyForged >= header.height) {
+			return;
```

Comment:
eslint(consistent-return) :D Seriously why eslint is not complaining as much as i do :D

Reply:
Also please avoid undefined returns.

Reply:
Because you are more tedious reviewer than eslint 😄

Reply:
I would change that, but that change was not part of this PR. Its just the issue Git didn't tracked the file rename very well.

Reply:
Let's ignore this comment too.